File: cpp/randAdapLU.cpp
#include "rid.hpp"
#include "util.hpp"

#include <iostream>
#include <assert.h>


void RandAdapLUPP(const Mat &A, double tol, int blk) {
  /*
   * Compute (row) ID decomposition
   */
  
  int m = A.rows();
  int n = A.cols();

  assert( n > blk );

  Eigen::PermutationMatrix<Eigen::Dynamic, Eigen::Dynamic> P( m );
  P.setIdentity();

  int p = std::min(m, n);
  Mat L = Mat::Zero( m, p );
  Mat U = Mat::Zero( p, n );

  Mat Y = RandColSketch(A, blk);
  Mat E = Y;

  int nb = std::ceil( p/blk );
  assert( p%blk == 0 ); // for current implementation; will be removed
  
  for (int i=0; i<nb; i++) {
    int k = i*blk; // number of processed rows/columns
 
    MKL_INT b = E.cols(); // current block size
    MKL_INT ipiv[ b ]; // 1-based fortran style integers
    assert( int(b)==blk );
    
    MKL_INT a = E.rows();
    MKL_INT info;
    dgetrf( &a, &b, E.data(), &a, ipiv, &info );
    assert( info==0 );
    assert( E.rows() >= E.cols() );
    

    // U_1^t
    U.block(k, k, b, b) = E.topRows(b).triangularView<Eigen::Upper>();
    
    // global permutation (accumulation of local results)
    for (int j=0; j<b; j++)
      P.applyTranspositionOnTheLeft( k+j, k+ipiv[j]-1 );

    Mat L1 = E.topRows(b).triangularView<Eigen::UnitLower>();
    Mat L2 = E.bottomRows(a-b);

    // L matrix
    L.block(k, k, b, b) = L1;
    L.block(k+b, k, a-b, b) = L2;

    // local permutation of current iteration
    Eigen::PermutationMatrix<Eigen::Dynamic, Eigen::Dynamic> Phat( a );
    Phat.setIdentity();
    for (int j=0; j<b; j++)
      Phat.applyTranspositionOnTheLeft( j, ipiv[j]-1 );

    if (i>0) L.bottomLeftCorner( a, k ) = Phat * L.bottomLeftCorner( a, k );


    // next panel
    if (i == nb-1) break;
      
    b = std::min(blk, p-(nb-1)*blk); // handle last panel
    Y = P * RandColSketch(A, b);

    k += blk; // number of processed rows/columns
    U.block(0, k, k, b ) = 
      L.topLeftCorner( k,k ).triangularView<Eigen::Lower>()
      .solve( Y.topRows( k ) );

    E = Y.bottomRows( m-k ) - 
      L.bottomLeftCorner( m-k, k ) * U.block( 0, k, k, b );

    double eSchur = E.norm();
    std::cout<<"Norm of Schur complement: "<<eSchur<<std::endl;
    if ( eSchur < tol ) break;
  }
}

File: matlab/RandAdapLU.m

function [sk, rd, T, flops] = RandAdapLU(A, blk, tol)

[m, n] = size(A);
X = zeros(m, n);
P = 1:m;

Y = A*RandMat(n,blk);
flops = 2*m*n*blk;

p = min(m, n);
nb = ceil(p/blk);
err = nan(nb,1);
err(1) = norm(Y, 'fro');

for i=0:nb-1
    k = i*blk;
    if i < nb-1
        b = blk;
    else
        b = p-(nb-1)*blk;
    end
    
    % inplace LU
    [L,~,phat] = lu( Y(k+1:end,1:b), 'vector' );
    X(k+1:end,k+1:k+b) = L;
    flops = flops + 2*(m-k)*b*b/3;
    
    % global permutation
    tmp1 = P(k+1:end);
    P(k+1:end) = tmp1(phat);
    
    % apply local permutation
    if i>0
        tmp2 = X(k+1:end,1:k);
        X(k+1:end,1:k) = tmp2(phat,:);
    end
    
    if i==nb-1, break, end
    
    b = min(blk, p-(nb-1)*blk);
    Y = A*RandMat(n,b);
    Y = Y(P,:);
    flops = flops + 2*m*n*b;
    
    % Schur complement
    k = k + blk;
    Y(k:end,:) = Y(k:end,:) - X(k:end,1:k) * (X(1:k,1:k) \ Y(1:k,:));
    flops = flops + k*k*b + 2*(m-k)*k*b;
    
    err(i+2) = norm(Y(k:end,:), 'fro');
    %fprintf("Norm of Schur complement: %d\n", eSchur);
    if err(i+2) < tol, break, end
end

r = k;
sk = P(1:r);
rd = P(r+1:end);
T = X(r+1:end,1:r)/X(1:r,1:r);
flops = flops + r*r*(m-r);

end